### PR TITLE
Feedback designated initializer changes + `Feedback(source:as:)`.

### PR DESCRIPTION
**Pull request by andersio**
<li>Introduce `Feedback(source:as:)`, which maps all values from the source to be specific feedback loop events.</li>

<li>The public designated initializer passes now accepts `SignalProducer<Never, Never>` created by `enqueue(to:)`. It is no longer necessary to `start()` the producer manually, when writing custom feedback variants.
    <table><tr><th>Before (Async API)</td><th>Before</th></tr><tr><td><pre lang="swift">
    Feedback { scheduler, state in
        state
            .diff()
            .flatMap(.concat, processDiff)
            .observe(on: scheduler)
    }</pre></td><td><pre lang="swift">
    Feedback { state, output in
        state
            .diff()
            .flatMap(.concat, processDiff)
            .enqueue(to: output)
            .start()
    }</pre></td></tr></table>
    <table><tr><th>After</th></tr><tr>
<td><pre lang="swift">
    Feedback { state, output in
        state
            .diff()
            .flatMap(.concat, processDiff)
            .enqueue(to: output)
    }</pre></td></tr>
</table></li>

<li>Deprecate `Feedback.custom` and the Disposable-expecting `Feedback.init`.</li>

<li>The setup closure `Feedback.init` now gives you a state producer would replay the latest state whenever you start it. The practical constraints are documented in the initializer docs.</li>

### CODE REVIEW

File: Loop/FeedbackLoop.swift [hunk 1]
```diff
@@ -52,31 +52,51 @@ extension FeedbackLoop {
...
+        ///
+        /// Events must be explicitly enqueued using `SignalProducer.enqueue(to:)` with the `FeedbackEventConsumer`
+        /// provided to the setup closure. `enqueue(to:)` respects producer cancellation and removes outstanding events
+        /// from the loop internal event queue.
+        ///
+        /// This is useful if you wish to discard events when the state changes in certain ways. For example,
+        /// `Feedback(skippingRepeated:effects:)` enqueues events inside `flatMap(.latest)`, so that unprocessed events
+        /// are automatically removed when the inner producer has switched.
+        ///
+        /// - important: The `state` producer provided to the setup closure **does not** replay the current state.
```
Comment (sergdort) [inline Loop/FeedbackLoop.swift]: Hm i wondering why do they use [replay subject](https://github.com/NoTests/RxFeedback.swift/blob/master/Sources/RxFeedback/ObservableType%2BRxFeedback.swift#L35) in Rx 🤔

Comment (andersio) [reply Loop/FeedbackLoop.swift]: That's interesting. I can look at whether we can make it provide replay-1 semantic.

File: Loop/FeedbackLoop.swift [hunk 2]
```diff
@@ -71,7 +59,58 @@ extension FeedbackLoop {
...
+        /// You can however expect it to be eventually consistent:
+        /// ```swift
+        /// Feedback { state, output in
+        ///     state
+        ///        .filter { $0.apples.isEmpty == false }
+        ///        .map(value: Event.eatAllApples)
+        ///        .take(first: 1)
+        ///        .concat(
+        ///            state
+        ///                .filter { $0.apples.isEmpty }
```
Comment (andersio) [inline Loop/FeedbackLoop.swift]:
```suggestion
        ///                .filter { $0.apples.isEmpty } // ℹ️ Watching specifically for the ideal state.
```

**Review by sergdort — APPROVED** [review]

**Comment by RuiAAPeres** [comment]
```swift
Feedback { state, output in
    state
        .diff()
        .flatMap(.concat, processDiff)
        .enqueue(to: output)
}
```

What's `output` and why does it need to enqueued? I am curious to understand about the motivations behind this PR and how does it help consumers of this API.

**Comment by sergdort** [comment]
@RuiAAPeres I believe `output` was already there, it's a thing that process values emitted from the producer

**Comment by RuiAAPeres** [comment]
My frame of though is set on the previous RAF approach using `Property` as the starting point. A typical Feedback, under that system, would return a `SignalProducer` where its `SignalProducer.Value` would be an `Event`. There were ofc many variants and among them:

```swift
    public init(events: @escaping (Scheduler, Signal<State, Never>) -> Signal<Event, Never>) {

public init<U, Effect: SignalProducerConvertible>(
        deriving transform: @escaping (Signal<State, Never>) -> Signal<U, Never>,
        effects: @escaping (U) -> Effect
```

~The new approach, seems to use the `output` as the escape hatch, where one funnels whatever is processed  into said `output` via `.enqueue`~.  This begs two questions:

1. Why this design versus the previous one? Is this tied to the new reducer's signature? 
2. What would happen if for some reason I forgot to `.enqueue(to: output)`?

**Comment by sergdort** [comment]
> What would happen if for some reason I forgot to .enqueue(to: output)?

It will be gone, that's why we have a lot of convenience initialisers that hide this complexity
Version with `output` is needed for advance usage.

**Comment by RuiAAPeres** [comment]
@sergdort thank you for your answer. Could you elaborate on these advance cases? I had a look at the header of the method, but I couldn't understand the apples example.

**Comment by andersio** [comment]
@RuiAAPeres 

> What would happen if for some reason I forgot to `.enqueue(to: output)`?

Your events would not hit the reducer at all.

The signature requires `SignalProducer<Never, Never>`, so in most cases the type system would warn you about type mismatching.

We can also have `enqueue(to:)` to return a producer with a phantom error type, e.g. `SignalProducer<Never, EnqueueError>`, so that prevents even "completables" from accidentially satisfying the signature.

**Comment by andersio** [comment]
@RuiAAPeres 

> Why this design versus the previous one? Is this tied to the new reducer's signature?

Of the two variants you mentioned, only the first one is fundementally changed. The second one is renamed to `Feedback(compacting:effects:)`.

RAF was previously using an asynchronous DispatchQueue to break deadlocks that would otherwise happen, because RAS does not support recursive value delivery. There is _always_ an implicit requirement, if unfortunately undocumented, for all feedbacks to enqueue their events to a specific `Scheduler` (backed by the said DispatchQueue) using `observe(on:)`.

We have provided higher-order Feedback variants that has compliant implementations, so the only place that it matters was the designated initializer accepting `(Scheduler, Signal<State, Never> -> Signal<Event, _Never>`.

In Loop, the asynchronous `DispatchQueue` is replaced with a synchronous queue-drain model — _queue_ here refers to _queue_ the data structure, not the libdispatch task queue. We cannot retrofit the support in without changing the designated initializer, because:

1. We need to process the events inside a critical section.

2. We need to queue the event if the lock is contended (usually due to recursion).

3. Events CANNOT escape the critical section.

   The `DispatchQueue` based model can tolerate events accidentially escaping the designated scheduler after `observe(on:)`, because we only need the async scheduling as a deadlock breaker. The work is free to run on any arbitrary `DispatchQueue`, thanks to RAS being thread safe by default.

    ```swift
    // Deprecated async API
    Feedback { scheduler, state in
        state
            .doSomething()
            .observe(on: scheduler)
            .doSomeOtherThings()    // tolerable in RAF
            .doMoreThings()         // tolerable in RAF
    }
    ```

    In the new queue-drain model, we need to start a critical section **and** append any emitted event to the internal event queue immediately. This is an implementation challenge, because the `Scheduler` abstraction cares only about _action closures_, with no visibility to values & terminations.

    Most importantly, even if we worked this around, the fact that you can apply any aribtrary operator — including all async ones — after `observe(on:)` breaks all the constraints on which the synchronous queue-drain model needs to operate. That is unless we add more ad-hoc rules — like the one above about applying `observe(on:)` — but these are brittle. In other words, these requirements are intrinsically unenforcable as is, since the premise of the pub-sub pattern is that subscribers are opaque to publishers.

4. We want to maintain the auto-cancellation semantics of many higher-order Feedback variants we have.

     The event enqueuing/processing is required to be adaptable to different part of the SignalProducer graph. For example, instead of being applied to the composition tail, we use it in `flatMap(.latest)` in many cases:

    ```swift
    // Deprecated async API
    Feedback { scheduler, state in
        state
            .flatMap(.latest) { state -> SignalProducer<Event, Never> in
                transform(state).producer
                    .observe(on: output)
            }
    }
    ```

    This means we need an explicit notion of "enqueue all events here, but dequeue them if I got cancelled" that can be adapted to any reasonable location. `observe(on:)` serves this double duty in RAF, since it supports cancelling scheduled tasks.


So that leads us to the introduction of the duo — the `enqueue(to:)` operator _and_ the new designated initializer signature.

**Pull request merged**